apps/emqx_durable_storage/src/emqx_ds_storage_layer_kv.erl

```diff
@@ -0,0 +1,180 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+-module(emqx_ds_storage_layer_kv).
+
+%% API:
+-export([prepare_kv_tx/5, commit_batch/4, lookup_kv/3]).
+
+%% internal exports:
+-export([]).
+
+-export_type([]).
+
+-include_lib("snabbkaffe/include/trace.hrl").
+-include("emqx_ds.hrl").
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+-type cooked_tx() :: term().
+
+-callback create(
+    emqx_ds_storage_layer:dbshard(),
+    rocksdb:db_handle(),
+    emqx_ds:generation(),
+    Options :: map(),
+    emqx_ds_storage_layer:generation_data() | undefined,
+    emqx_ds:db_opts()
+) ->
+    {_Schema, emqx_ds_storage_layer:cf_refs()}.
+
+%% Open the existing schema
+-callback open(
+    emqx_ds_storage_layer:dbshard(),
+    rocksdb:db_handle(),
+    emqx_ds:generation(),
+    emqx_ds_storage_layer:cf_refs(),
+    _Schema
+) ->
+    emqx_ds_storage_layer:generation_data().
+
+%% Delete the schema and data
+-callback drop(
+    emqx_ds_storage_layer:dbshard(),
+    rocksdb:db_handle(),
+    emqx_ds:generation(),
+    emqx_ds_storage_layer:cf_refs(),
+    emqx_ds_storage_layer:generation_data()
+) ->
+    ok | {error, _Reason}.
+
+-callback prepare_kv_tx(
+    emqx_ds_storage_layer:dbshard(),
+    emqx_ds_storage_layer:generation_data(),
+    emqx_ds:tx_serial(),
+    emqx_ds:kv_tx_ops(),
+    emqx_ds_storage_layer:batch_prepare_opts()
+) ->
+    {ok, cooked_tx()} | emqx_ds:error(_).
+
+-callback commit_batch(
+    emqx_ds_storage_layer:dbshard(),
+    emqx_ds_storage_layer:generation_data(),
+    [cooked_tx()],
+    emqx_ds_storage_layer:batch_store_opts()
+) -> ok | emqx_ds:error(_).
+
+-callback get_streams(
+    emqx_ds_storage_layer:dbshard(),
+    emqx_ds_storage_layer:generation_data(),
+    emqx_ds:topic_filter(),
+    _Time
+) ->
+    [_Stream].
+
+-callback make_iterator(
+    emqx_ds_storage_layer:dbshard(),
+    emqx_ds_storage_layer:generation_data(),
+    _Stream,
+    emqx_ds:topic_filter(),
+    beginning | emqx_ds:message_key()
+) ->
+    emqx_ds:make_iterator_result(_Iterator).
+
+-callback lookup(
+    emqx_ds_storage_layer:dbshard(),
+    emqx_ds_storage_layer:generation_data(),
+    emqx_ds:topic()
+) ->
+    {ok, binary()} | undefined.
+
+-callback next(
+    emqx_ds_storage_layer:dbshard(),
+    emqx_ds_storage_layer:generation_data(),
+    Iter,
+    pos_integer(),
+    %% FIXME: remove
```

Comment:
Currently calls to kv layouts are still dispatched via emqx_ds_storage_layer and carry some extra parameters. Fixing it is not super trivial and not very useful, except from aesthetics perspective, so there are some useless parameters.